### PR TITLE
PorousFlow: hystersis core doco added

### PR DESCRIPTION
**Pull request by WilkAndy**
Refs #16049

### CODE REVIEW

File: modules/porous_flow/doc/content/bib/porous_flow.bib [hunk 1]
```diff
@@ -487,3 +487,46 @@ @book{javendel
...
+year = {2007},
+pages = {1768--1781},
+title = {Modeling geologic storage of carbon dioxide: Comparison of non-hysteretic and hysteretic characteristic curves},
+volume = {48},
+journal = {Energy Conservation and Management},
+doi = {10.1016/j.enconman.2007.01.022}
+}
+
+@techreport{doughty2008,
+  title={User's guide for hysteretic capillary pressure and relative permeability functions in {iTOUGH2}, Rev3, Draft, March 2008},
```
Comment (cpgr) [inline modules/porous_flow/doc/content/bib/porous_flow.bib:502]: The March renders with a lowercase m on the website

File: modules/porous_flow/doc/content/bib/porous_flow.bib [hunk 2]
```diff
@@ -487,3 +487,46 @@ @book{javendel
...
+title = {Comparison of soil-water hysteresis models},
+volume = {75},
+journal = {Journal of Hydrology}
+}
+
+@article{niemi1988,
+author = {A. Niemi and G. S. Bodvarsson},
+year = {1988},
+pages = {277--291},
+title = {Preliminary capillary hysteresis simulations in fractured rocks, Yucca Mountain, Nevada},
```
Comment (cpgr) [inline modules/porous_flow/doc/content/bib/porous_flow.bib:528]: Same with yucca mountain, nevada

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
+# Hysteresis in PorousFlow
+
+Hysteretic capillary pressure and relative permeability functions are implemented in PorousFlow.  Hysteresis means that capillary pressure and relative permeability depend on the *history* as well as the current state.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:3]: Maybe the *saturation history*?

**Review by cpgr — COMMENTED** [review]
Apart from these minor comments, thanks for the thorough documentation Andy

**Comment by cpgr** [comment]
I'll add my comments for the other PR now too.

File: modules/porous_flow/doc/content/bib/porous_flow.bib (hunk 2 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/bib/porous_flow.bib:528]: done

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md (hunk 3 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:3]: thanks, done

File: modules/porous_flow/doc/content/bib/porous_flow.bib (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/bib/porous_flow.bib:502]: Thanks

**Comment by WilkAndy** [comment]
Hey @cpgr, this is done now.  I addressed your comments above, and added stuff about relative permeability.  Could you read the relperm stuff?  (The cap pressure stuff is the same)

**Comment by cpgr** [comment]
> Hey @cpgr, this is done now. I addressed your comments above, and added stuff about relative permeability. Could you read the relperm stuff? (The cap pressure stuff is the same)

yep. I'll take a look shortly.

**Comment by WilkAndy** [comment]
Ooops, @cpgr - i just realised that i didn't push all the hysteresis relperm doco up yesterday.  i've just done that now and hopefully all the tests will pass - i'll send you the link for the doco when that happens.

**Comment by WilkAndy** [comment]
The hysteresis doco, this time with the relperm stuff included, is at

https://mooseframework.inl.gov/docs/PRs/16057/site/modules/porous_flow/hysteresis.html

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+\begin{equation}
+\log P_{c} = \log P_{c}^{w} + (s - \mathrm{TP}_{1}) \frac{\log P_{c}^{d} - \log P_{c}^{w}}{\mathrm{TP}_{0} - \mathrm{TP}_{1}} \ .
+\end{equation}
+
+This produces continuous capillary-pressure functions as shown in [histeretic_order_fig].
+
+
+## Relative permeabilities
+
+Hysteresis is defined for two-phase systems only.  The water and gas relative permeability functions are both hysteretic.  Only the drying and first-order wetting curves are defined.  This means that if the system dries (following the drying curve) and then wets (following the first-order wetting curve) and then subsequently dries, the system will move along the first-order wetting curve until the turning point is reached, when it will move along the drying durve.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:174]: Typo (last word) - durve!

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+For reference, the zeroth-order (drying) are given by [relperm.eqns] but because $S_{gr}^{\Delta} = 0$ in this case, $\bar{S}_{gt} = 0$, and the functions reduce to
+\begin{equation}
+\begin{aligned}
+\label{relperm.eqns.0}
+k_{r,l}(S_{l}) & = \sqrt{\bar{S}_{l}} \left[1 - \left( 1- \bar{S}_{l}^{1/m} \right)^{m}\right]^{2} \ , \\
+k_{r,g}(S_{l}) & = k_{r,g}^{max} \left(1 - \bar{S}_{l}\right)^{\gamma} \left(1 - \bar{S}_{l}^{1/m} \right)^{2m} \ .
+\end{aligned}
+\end{equation}
+
+Assuming the relative-permeability values are constant outside the well-defined domain results in curves of the form shown in [histeretic_krel_unextended_fig].
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:212]: No hyphen?

Comment (cpgr) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:212]: In relative permeability, I mean.

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+   - $S_{\mathrm{small}} = r(1 - S_{gr}^{\Delta})$, where $r$ is a (dimensionless) user-input satisfying $0 < r < 1$, and optimally $r$ should be close to 1, for example $r=0.9$.
+   - $S_{\mathrm{big}} = 1 - S_{gr}^{\Delta} / 2$.  Note that [!citet](doughty2008) defines $S_{\mathrm{big}} = (2 - r)(1 - S_{gr}^{\Delta})$, but this can be greater than 1 and can result in a poor modification.
+
+ Then:
+
+   - for $S<S_{\mathrm{small}}$ the liquid wetting curve, given in [relperm.eqns], is used
+   - for $S>S_{\mathrm{big}}$ the liquid drying curve, given in [relperm.eqns.0], is used
+   - otherwise, a cubic spline interpolates between these two, with parameters chosen so that the final result is continuous and has a continuous derivative.
+
+The result is illustrated in [histeretic_krel_extended_fig], using a cubic extension for the gas relative-permeability curve, and $r = 0.9$.
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:236]: No hyphen?

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+!media media/porous_flow/hysteretic_krel_extended.png caption=Extended and modified relative permeability curves, using a cubic extension for the gas relative-permeability curve, and $r$ being 0.9.  id=histeretic_krel_extended_fig
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:238]: And here

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+### Examples
+
+!media media/porous_flow/hysteretic_krel_example_1.png caption=Example hysteretic relative-permeability functions.  The system initialises at full saturation, dries to saturation 0.5 (greater than $S_{l, r}$) and then wets back to saturation 1.  id=histeretic_krel_example_1
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:242]: and here

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+!media media/porous_flow/hysteretic_krel_example_2.png caption=Example hysteretic relative-permeability functions.  The system initialises at full saturation, dries to saturation 0.1 (less than $S_{l, r}$) and then wets back to saturation 1.  In the extended region ($S_{l} \leq S_{l, r}$), the drying curve equals the wetting curve.  id=histeretic_krel_example_2
```
Comment (cpgr) [inline modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:244]: and finally here

**Review by cpgr — DISMISSED** [review]
Looks great Andy.

File: modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md (hunk 4 above)
Comment (WilkAndy) [reply modules/porous_flow/doc/content/modules/porous_flow/hysteresis.md:174]: eagle eyes!

**Comment by WilkAndy** [comment]
I believe this is ready for merging, @aeslaughter

**Review by aeslaughter — APPROVED** [review]

**Comment by WilkAndy** [comment]
Wow, that was quick, thanks @aeslaughter

**Pull request merged**